### ISSUE TITLE
Swap units are deactivated too early

### ISSUE THREAD

**Issue by fsateler** [systemd/systemd#1902]
Forwarding this from downstream

> On a normal shutdown or reboot, systemd disables swap before stopping
> any other processes.  I see something like this:
> 
>  Nov 15 14:50:53 systemd[1]: Started Synchronise Hardware Clock to System Clock.
>  Nov 15 14:50:53 systemd[1]: Deactivating swap /dev/disk/by-uuid/bffdb4e8-...
>  Nov 15 14:50:53 systemd[1]: Deactivating swap /dev/disk/by-uuid/bffdb4e8-...
>  Nov 15 14:50:53 systemd[1]: Deactivating swap /dev/disk/by-uuid/bffdb4e8-...
>  Nov 15 14:50:53 systemd[1]: Stopping system-ifup.slice.
>  Nov 15 14:50:53 systemd[1]: Removed slice system-ifup.slice.
>  Nov 15 14:50:53 systemd[1]: Stopping <all other services>
> 
> Which is fairly undesirable, because processes wanting to do a clean
> shutdown may need to allocate memory, and they are now competing for
> it with the thundering herd of all the processes that have just been
> signalled to shutdown simultaneously.

Maybe swap units (with `DefaultDependencies=yes`) should have some sort of ordering so that they are shut down as late as possible, to prevent OOM errors during shutdown.

I can confirm with 227 that swap units are deactivated very early.

Bug-Debian: http://bugs.debian.org/805133

**Comment by fsateler** [comment]
More information from my own system. For some reason, the swap units are not aliases to each other, and that results in incorrect dependencies for 2 of the 3 versions. This also would explain why swaps are deactivated more than once at shutdown.

```
% swapon -s                                                                                                          
Filename                                Type            Size    Used    Priority
/dev/sda6                               partition       2309116 621808  -1
% systemctl show -pId,Names,Before dev-disk-by\\x2duuid-df3782e5\\x2d7997\\x2d48e5\\x2da501\\x2dcbd7bfe96e91.swap dev-sda6.swap dev-disk-by\\x2did-ata\\x2dSamsung_SSD_850_EVO_250GB_S21NNSAG462820E\\x2dpart6.swap
Id=dev-disk-by\x5cx2duuid-df3782e5\x5cx2d7997\x5cx2d48e5\x5cx2da501\x5cx2dcbd7bfe96e91.swap
Names=dev-disk-by\x5cx2duuid-df3782e5\x5cx2d7997\x5cx2d48e5\x5cx2da501\x5cx2dcbd7bfe96e91.swap
Before=umount.target swap.target

Id=dev-sda6.swap
Names=dev-sda6.swap
Before=umount.target

Id=dev-disk-by\x5cx2did-ata\x5cx2dSamsung_SSD_850_EVO_250GB_S21NNSAG462820E\x5cx2dpart6.swap
Names=dev-disk-by\x5cx2did-ata\x5cx2dSamsung_SSD_850_EVO_250GB_S21NNSAG462820E\x5cx2dpart6.swap
Before=umount.target
% systemctl show -pId,Names dev-disk-by\\x2duuid-df3782e5\\x2d7997\\x2d48e5\\x2da501\\x2dcbd7bfe96e91.device dev-sda6.device dev-disk-by\\x2did-ata\\x2dSamsung_SSD_850_EVO_250GB_S21NNSAG462820E\\x2dpart6.device
Id=dev-disk-by\x5cx2duuid-df3782e5\x5cx2d7997\x5cx2d48e5\x5cx2da501\x5cx2dcbd7bfe96e91.device
Names=dev-disk-by\x5cx2duuid-df3782e5\x5cx2d7997\x5cx2d48e5\x5cx2da501\x5cx2dcbd7bfe96e91.device

Id=dev-sda6.device
Names=dev-sda6.device

Id=dev-disk-by\x5cx2did-ata\x5cx2dSamsung_SSD_850_EVO_250GB_S21NNSAG462820E\x5cx2dpart6.device
Names=dev-disk-by\x5cx2did-ata\x5cx2dSamsung_SSD_850_EVO_250GB_S21NNSAG462820E\x5cx2dpart6.device

```

My relevant fstab line is 

```
UUID=df3782e5-7997-48e5-a501-cbd7bfe96e91 none            swap    sw              0       0
```

Interesting is the fact that the only swap unit I can `systemctl cat` is the by-uuid version. The others give me "No files found for $unit.swap".

```
# /run/systemd/generator/dev-disk-by\x2duuid-df3782e5\x2d7997\x2d48e5\x2da501\x2dcbd7bfe96e91.swap
# Automatically generated by systemd-fstab-generator

[Unit]
SourcePath=/etc/fstab
Documentation=man:fstab(5) man:systemd-fstab-generator(8)

[Swap]
What=/dev/disk/by-uuid/df3782e5-7997-48e5-a501-cbd7bfe96e91
Options=sw
```

**Comment by fsateler** [comment]
Adding an add-in snippet to swap.target (in /etc) adding all aliases to After, enforces the correct ordering. So it seems that the problem is in fact that the duplicated units are not preserving the ordering constraints.

**Comment by poettering** [comment]
Should be fixed by #1997. Closing. Feel free to reopen, if the issue is not fixed.

**Comment by vsh01** [comment]
Hi, I used an adapted patch from #1997 for systemd 215-17+deb8u2 in debian jessie.
```udiff
Index: systemd-215/src/core/swap.c
===================================================================
--- systemd-215.orig/src/core/swap.c
+++ systemd-215/src/core/swap.c
@@ -224,6 +224,13 @@ static int swap_add_default_dependencies
         if (detect_container(NULL) > 0)
                 return 0;
 
+        /* swap units generated for the swap dev links are missing the
+         * ordering dep against the swap target. */
+        r = unit_add_dependency_by_name(UNIT(s), UNIT_BEFORE, SPECIAL_SWAP_TARGET, NULL, true);
+        if (r < 0)
+            return r;
+
+
         r = unit_add_two_dependencies_by_name(UNIT(s), UNIT_BEFORE, UNIT_CONFLICTS, SPECIAL_UMOUNT_TARGET, NULL, true);
         if (r < 0)
                 return r;
```
But the issue reproduced.
